File: app-iOS/Inventory Manager/Models/Location.swift
//
//  Location.swift
//  InventoryManager
//
//  Created by Ryan Mackin on 12/11/20.
//

import Foundation
import MapKit

struct Location: Codable {
    var timestamp: String
    var status: String
    
    // Address
    var street: String?
    var city: String?
    var state: String?
    var zip: String?
    var country: String?
    
    var address: String {
        if let street = street, let city = city, let state = state, let zip = zip, let country = country {
            return String("\(street), \(city), \(state) \(zip), \(country)")
        } else {
            return "Unknown Address"
        }
    }
    
    // Geolocation
    var latitude: String?
    var longitude: String?

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(
            latitude: Double(latitude ?? "0.0") ?? 0.0,
            longitude: Double(longitude ?? "0.0") ?? 0.0
        )
    }
}

File: server/models/inventory/location.js
module.exports = (sequelize, DataTypes) => {
    const Location = sequelize.define('Location', {
        id: { type: DataTypes.INTEGER, autoIncrement: true, primaryKey: true },

        timestamp: { type: DataTypes.DATE, allowNull: false, defaultValue: DataTypes.NOW },

        street: { type: DataTypes.STRING, allowNull: true },
        city: { type: DataTypes.STRING, allowNull: true },
        state: { type: DataTypes.STRING, allowNull: true },
        zip: { type: DataTypes.STRING, allowNull: true },
        country: { type: DataTypes.STRING, allowNull: true },

        latitude: { type: DataTypes.STRING, allowNull: true },
        longitude: { type: DataTypes.STRING, allowNull: true },

        status: { type: DataTypes.STRING, allowNull: true },

        createdAt: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
        updatedAt: { type: DataTypes.DATE, defaultValue: DataTypes.NOW }
    }, {
        tableName: 'Locations'
    })

    return Location
}